tests/compile/test_builders.py

```diff
@@ -115,10 +115,14 @@ def test_grad_grad(self, cls_ofg):
         x, y, z = matrices("xyz")
         e = x + y * z
         op = cls_ofg([x, y, z], [e])
+
         f = op(x, y, z)
-        f = f - grad(pt_sum(f), y)
-        f = f - grad(pt_sum(f), y)
-        fn = function([x, y, z], f)
+        g = grad(pt_sum(f), y)
```

Comment:
Can you add a new test for the error and leave this one unchanged? Or did this start feeling? Looks like it shouldn't have?

Reply:
Start feeling?

Reply:
Hmm why? It was working before, may need to investigate or is it obvious?

Reply:
It merits investigation. But it might be that the test was always flakey and now we're catching it. The duplicated input is one of the intermediate `f` representations -- the true inputs x,y,z aren't the problem. I assume this is something to do with how the grad function works?

Reply:
Intermediate variables can be repeated just fine

Reply:
Perhaps the grad OFG is the one with duplicated input variables?

Reply:
Trying to dig into this, the error is raised from a `NominalTensorVariable` but this phrase occurs nowhere in the codebase. Any idea what it is? Full traceback hidden below.

<details><summary>Details</summary>
<p>

```
tests/compile/test_builders.py:110 (TestOpFromGraph.test_grad_grad[cls_ofg1])
self = <tests.compile.test_builders.TestOpFromGraph object at 0x7f0b4bd545b0>
cls_ofg = functools.partial(<class 'pytensor.compile.builders.OpFromGraph'>, inline=True)

    @pytest.mark.parametrize(
        "cls_ofg", [OpFromGraph, partial(OpFromGraph, inline=True)]
    )
    def test_grad_grad(self, cls_ofg):
        x, y, z = matrices("xyz")
        e = x + y * z
        op = cls_ofg([x, y, z], [e])
        f = op(x, y, z)
        g1 = grad(pt_sum(f), y)
        f = f - grad(pt_sum(f), y)
>       f = f - grad(pt_sum(f), y)

test_builders.py:121: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../pytensor/gradient.py:505: in grad
    var_to_app_to_idx = _populate_var_to_app_to_idx(outputs, _wrt, consider_constant)
../../pytensor/gradient.py:949: in _populate_var_to_app_to_idx
    account_for(output)
../../pytensor/gradient.py:945: in account_for
    account_for(ipt)
../../pytensor/gradient.py:945: in account_for
    account_for(ipt)
../../pytensor/gradient.py:921: in account_for
    connection_pattern = _node_to_pattern(app)
../../pytensor/gradient.py:800: in _node_to_pattern
    connection_pattern = node.op.connection_pattern(node)
../../pytensor/compile/builders.py:893: in connection_pattern
    lop_op = self.get_lop_op()
../../pytensor/compile/builders.py:742: in get_lop_op
    self._recompute_lop_op()
../../pytensor/configparser.py:47: in res
    return f(*args, **kwargs)
../../pytensor/compile/builders.py:610: in _recompute_lop_op
    self._lop_op = type(self)(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = OpFromGraph()
inputs = [NominalTensorVariable(0, TensorType(float64, shape=(None, None))), NominalTensorVariable(1, TensorType(float64, shape...le(4, TensorType(float64, shape=(None, None))), NominalTensorVariable(4, TensorType(float64, shape=(None, None))), ...]
outputs = [Second.0, Mul.0, Mul.0, Second.0, <Matrix(float64, shape=(?, ?))>]
inline = True, lop_overrides = 'default', grad_overrides = 'default'
rop_overrides = 'default', connection_pattern = None, strict = False
name = None, kwargs = {'on_unused_input': 'ignore'}

    def __init__(
        self,
        inputs: list[Variable],
        outputs: list[Variable],
        *,
        inline: bool = False,
        lop_overrides: str = "default",
        grad_overrides: str = "default",
        rop_overrides: str = "default",
        connection_pattern: list[list[bool]] | None = None,
        strict: bool = False,
        name: str | None = None,
        **kwargs,
    ):
        """
        Parameters
        ----------
        inputs
            The inputs to the graph.
        outputs
            The outputs to the graph.
        inline
            Defaults to ``False``
    
            ``True`` : Cause the :class:`Op`'s original graph being used during
            compilation, the :class:`Op` will not be visible in the compiled
            graph but rather its internal graph.
    
            ``False`` : will use a pre-compiled function inside.
        grad_overrides
            Defaults to ``'default'``.
            This argument is mutually exclusive with ``lop_overrides``.
    
            ``'default'`` : Do not override, use default grad() result
    
            `OpFromGraph`: Override with another `OpFromGraph`, should
            accept inputs as the same order and types of ``inputs`` and ``output_grads``
            arguments as one would specify in :meth:`Op.grad`() method.
    
            `callable`: Should take two args: ``inputs`` and ``output_grads``.
            Each argument is expected to be a list of :class:`Variable `.
            Must return list of :class:`Variable `.
        lop_overrides
            Defaults to ``'default'``.
    
            This argument is mutually exclusive with ``grad_overrides``.
    
            These options are similar to the ``grad_overrides`` above, but for
            the :meth:`Op.L_op` method.
    
            ``'default'``: Do not override, use the default :meth:`Op.L_op` result
    
            `OpFromGraph`: Override with another `OpFromGraph`, should
            accept inputs as the same order and types of ``inputs``,
            ``outputs`` and ``output_grads`` arguments as one would specify in
            :meth:`Op.grad` method.
    
            `callable`: Should take three args: ``inputs``, ``outputs`` and ``output_grads``.
            Each argument is expected to be a list of :class:`Variable`.
            Must return list of :class:`Variable`.
    
            `NullType` instance: Treat as non-differentiable
            `DisconnectedType` instance: Treat as disconnected gradient,
            numerically gives zero
    
            ``list``: Each `OpFromGraph`/callable must return a single
            :class:`Variable`. Each list element corresponds to gradient of
            a specific input, length of list must be equal to number of inputs.
    
        rop_overrides
            One of ``{'default', OpFromGraph, callable, Variable}``.
    
            Defaults to ``'default'``.
    
            ``'default'``: Do not override, use the default :meth:`Op.R_op` result
    
            `OpFromGraph`: Override with another `OpFromGraph`, should
            accept inputs as the same order and types of ``inputs`` and ``eval_points``
            arguments as one would specify in :meth:`Op.R_op` method.
    
            `callable`: Should take two args: ``inputs`` and ``eval_points``.
            Each argument is expected to be a list of :class:`Variable`.  Must
            return list of :class:`Variable`.
    
            `NullType` instance: Treat as non-differentiable `DisconnectedType`
            instance: Treat as zero since `DisconnectedType` is not yet supported
            in :meth:`Op.R_op`.
    
            ``list``:
            Each :class:`OpFromGraph`/callable must return a single
            :class:`Variable <pytensor.graph.basic.Variable>`. Each list element
            corresponds to a specific output of :meth:`Op.R_op`, length of list
            must be equal to number of outputs.  connection_pattern If not
            ``None``, this will be used as the connection_pattern for this
            :class:`Op`.
        strict: bool, default False
            If true, it raises when any variables needed to compute the inner graph
            are not provided as explici inputs. This can only happen for graphs with
            shared variables.
        name
            A name for debugging purposes.
        kwargs
            Check :func:`pytensor.function` for more arguments, only works when not
            inline.
        """
    
        if len(inputs) != len(set(inputs)):
            var_counts = {var: inputs.count(var) for var in inputs}
            duplicated_inputs = [var for var, count in var_counts.items() if count > 1]
>           raise ValueError(
                f"There following variables were provided more than once as inputs to the OpFromGraph, resulting in an "
                f"invalid graph: {duplicated_inputs}. Use dummy variables or var.copy() to distinguish "
                f"variables when creating the OpFromGraph graph."
            )
E           ValueError: There following variables were provided more than once as inputs to the OpFromGraph, resulting in an invalid graph: [NominalTensorVariable(4, TensorType(float64, shape=(None, None)))]. Use dummy variables or var.copy() to distinguish variables when creating the OpFromGraph graph.

../../pytensor/compile/builders.py:407: ValueError
```

</p>
</details>

Reply:
Here's the simplest graph I could cook up that triggers the error:

```
a = pt.dmatrix('a')
op = pytensor.compile.builders.OpFromGraph([a], [a])
pt.grad(op(a).sum(), a)
```

Reply:
Right, the `L_op` creates a node that has as inputs: inputs + outputs + outputs_grads, in your case inputs and outputs are both `a`. That's masked by them passing the kwarg `on_unused_input="ignore"` internally when creating the `L_op` OFG.

In #723 I added a TODO comment that we could be more clever about the inputs/outputs and only define the restricted L_op we actually need and avoid the unused_input. In any case, for this PR it means we probably want to revert the check we added, or maybe make it exclusive on `kwargs.get("on_unused_input", False) != "ignore"`?

Reply:
The test doesn't have obvious double inputs though, what's going on there? Is it because the original `f` is considered input to both `grad(f.sum(), y)` and `grad(grad(f.sum(), y).sum(), y)` ?

Reply:
Let me see what are the duplicate inputs. In any case you proved we are too restrictive with the `OFG([a], [a])` example.

Reply:
The `L_op` of the first grad call looks like:

```python
pytensor.dprint(self.fgraph, print_fgraph_inputs=True)

→ *0-<Matrix(float64, shape=(?, ?))> [id A]
→ *1-<Matrix(float64, shape=(?, ?))> [id B]
→ *2-<Matrix(float64, shape=(?, ?))> [id C]
→ *3-<Matrix(float64, shape=(?, ?))> [id D]
→ *4-<Matrix(float64, shape=(?, ?))> [id E]
*4-<Matrix(float64, shape=(?, ?))> [id E]
Mul [id F] 0
 ├─ *4-<Matrix(float64, shape=(?, ?))> [id E]
 └─ *2-<Matrix(float64, shape=(?, ?))> [id C]
Mul [id G] 1
 ├─ *4-<Matrix(float64, shape=(?, ?))> [id E]
 └─ *1-<Matrix(float64, shape=(?, ?))> [id B]
```
You can see the last input `4*` is also the first output. So when the second L_op is called on the first one, it will try to build an `OFG` that takes as inputs `4*` twice, once from its inputs, and once from its outputs.

Reply:
I went with skipping the input check if on_unused_inputs == 'ignore'. I thought that was better than removing the error. I guess a real solution will be to check for outputs that are inputs when we build the OpFromGraph, and clone them if so?

Reply:
Agree, but let's do that in a separate PR. Then we can get rid of this hacky params as well